base/linalg/diagonal.jl

```diff
@@ -260,8 +262,10 @@ end
 @inline A_mul_Bc(D::Diagonal, rowvec::RowVector) = D*ctranspose(rowvec)
 
 conj(D::Diagonal) = Diagonal(conj(D.diag))
-transpose(D::Diagonal) = D
-ctranspose(D::Diagonal) = conj(D)
+transpose(D::Diagonal{<:Number}) = D
+transpose(D::Diagonal) = Diagonal(transpose.(D.diag))
+ctranspose(D::Diagonal{<:Number}) = conj(D)
+ctranspose(D::Diagonal) = Diagonal(ctranspose.(D.diag))
 
 diag(D::Diagonal) = D.diag
 trace(D::Diagonal) = sum(D.diag)
```

Comment:
`trace`, `det`, etc confuse me. Are these meant to be recursive? Should these only work for elements which have a ring structure, such as square matrices of the same size (and return such a matrix), or should it reflect any general block-diagonal structure (and return an actual scalar - i.e. `T` in `Diagona{Matrix{T}}`)? 

(@stevengj @StefanKarpinski this reflects exactly what I was trying to get to the bottom of in the long "Taking matrix transposes seriously" discussion - any help would be appreciated greatly)